### PR TITLE
ffmpeg_7: fix darwin build

### PR DESCRIPTION
**Pull request by jopejoe1**
## Description of changes

FFmpeg 7 on x86_64-darwin currently fails to compile with AV Foundation enabled, so disabling for x86_64-darwin on ffmpeg 7 or newer.
Also, xeve and xevd fail to compile with non GCC compilers, so disabling them on non gnu cc.

A Test for ffmpef_7-full on Darwin both architectures is currently still failing.

```
ast 10 log lines:
       > HOSTCC tests/audiogen.o
       > HOSTCC tests/videogen.o
       > COPY   tests/data/filtergraphs/anequalizer
       > COPY   tests/data/filtergraphs/firequalizer
       > COPY   tests/data/filtergraphs/compand
       > GEN    tests/data/hls-list.m3u8
       > make: *** [tests/fate/filter-audio.mak:209: tests/data/hls-list.m3u8] Error 134
       > make: *** Waiting for unfinished jobs....
       > GEN    tests/data/hls-list-append.m3u8
       > make: *** [tests/fate/filter-audio.mak:219: tests/data/hls-list-append.m3u8] Error 134
```

Full build log https://hydra.nixos.org/build/263289438/nixlog/1

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/xe/xevd/package.nix [hunk 1]
```diff
@@ -47,6 +47,6 @@ stdenv.mkDerivation (finalAttrs: {
     pkgConfigModules = [ "xevd" ];
     maintainers = with lib.maintainers; [ jopejoe1 ];
     platforms = lib.platforms.all;
-    broken = !stdenv.hostPlatform.isx86 || stdenv.hostPlatform.isDarwin;
+    broken = !stdenv.hostPlatform.isx86 || !stdenv.cc.isGNU;
```
Comment (Atemu) [inline pkgs/by-name/xe/xevd/package.nix:50]: A comment with the error and/or an upstream reference would be great.

File: pkgs/development/libraries/ffmpeg/generic.nix [hunk 2]
```diff
@@ -39,7 +39,7 @@
 , withAribcaption ? withFullDeps && lib.versionAtLeast version "6.1" # ARIB STD-B24 Caption Decoder/Renderer
 , withAss ? withHeadlessDeps && stdenv.hostPlatform == stdenv.buildPlatform # (Advanced) SubStation Alpha subtitle rendering
 , withAudioToolbox ? withHeadlessDeps && stdenv.isDarwin # Apple AudioToolbox
-, withAvFoundation ? withHeadlessDeps && stdenv.isDarwin # Apple AVFoundation framework
+, withAvFoundation ? withHeadlessDeps && stdenv.isDarwin && !(lib.versionAtLeast version "7" && stdenv.hostPlatform.isx86) # Apple AVFoundation framework
```
Comment (Atemu) [inline pkgs/development/libraries/ffmpeg/generic.nix:42]: Again, a comment briefly describing the current situation would be great.

**Review by Atemu — APPROVED** [review]
Diff LGTM but I can't test this myself.

If Darwin CI is green though, that'd be good enough for me.

**Review by JohnRTitor — APPROVED** [review]
Built on `aarch64-darwin`

**Comment by Atemu** [comment]
@JohnRTitor could you build on x86 via rosetta? That's what's being fixed here. aarch64-darwin was working already.

**Comment by DontEatOreo** [comment]
Result of `nixpkgs-review pr 322724` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Review by 4evy — APPROVED** [review]

**Comment by JohnRTitor** [comment]
@DontEatOreo I assume you cross compiled it on x86-darwin?

**Comment by DontEatOreo** [comment]
Result of `nixpkgs-review pr 322724` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>handbrake</li>
  </ul>
</details>
<details>
  <summary>7 packages failed to build:</summary>
  <ul>
    <li>ffmpeg_7-full</li>
    <li>ffmpeg_7-full.bin</li>
    <li>ffmpeg_7-full.data</li>
    <li>ffmpeg_7-full.dev</li>
    <li>ffmpeg_7-full.doc</li>
    <li>ffmpeg_7-full.lib</li>
    <li>ffmpeg_7-full.man</li>
  </ul>
</details>
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>ffmpeg_7</li>
    <li>ffmpeg_7-headless</li>
    <li>ffmpeg_7-headless.bin</li>
    <li>ffmpeg_7-headless.data</li>
    <li>ffmpeg_7-headless.dev</li>
    <li>ffmpeg_7-headless.doc</li>
    <li>ffmpeg_7-headless.lib</li>
    <li>ffmpeg_7-headless.man</li>
    <li>ffmpeg_7.bin</li>
    <li>ffmpeg_7.data</li>
    <li>ffmpeg_7.dev</li>
    <li>ffmpeg_7.doc</li>
    <li>ffmpeg_7.lib</li>
    <li>ffmpeg_7.man</li>
  </ul>
</details>

**Comment by Atemu** [comment]
Huh, full failed? Could you post the log?

@ofBorg build ffmpeg_7-full

**Comment by DontEatOreo** [comment]
> Huh, full failed? Could you post the log?
> 
> @ofborg build ffmpeg_7-full

https://termbin.com/tk9f

**Comment by jopejoe1** [comment]
The issue on ffmpeg_7-full on Darwin has something to do with ffmpeg not finding libquirc.

**Comment by Atemu** [comment]
The paste link is dead.

**Comment by emilazy** [comment]
If nobody figures this out before I finish dealing with FFmpeg 4, I’ll try to take a look at it then.

**Comment by jopejoe1** [comment]
Found the issue it seems libquirc was not linked correctly it was found during compile time but wen trying to run the build ffmpeg binary it failed to find it. 

For now i disabled as i do not know why it was linked incorrectly.

File: pkgs/by-name/xe/xevd/package.nix (hunk 1 above)
Comment (jopejoe1) [reply pkgs/by-name/xe/xevd/package.nix:51]: https://github.com/mpeg5/xeve/issues/117

and from reading the cmake file

File: pkgs/development/libraries/ffmpeg/generic.nix (hunk 2 above)
Comment (jopejoe1) [reply pkgs/development/libraries/ffmpeg/generic.nix:43]: Done

**Comment by JohnRTitor** [comment]
@ofborg build ffmpeg_7-full

**Comment by emilazy** [comment]
Building `ffmpeg_7-full` on native `x86_64-darwin`.

It seems dodgy that `quirc` contains a `.so` on a platform that uses `.dylib`. I don’t know if that’s the cause or not, but it looks like @toonn, who I know was working on this recently, got that fixed upstream: https://github.com/dlbeer/quirc/pull/144.

Perhaps we could consider including that patch or bumping to that commit and see if it fixes the issue?

**Comment by emilazy** [comment]
It does build!

**Comment by DontEatOreo** [comment]
Result of `nixpkgs-review pr 322724` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>handbrake</li>
  </ul>
</details>
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>ffmpeg_7-full</li>
    <li>ffmpeg_7-full.bin</li>
    <li>ffmpeg_7-full.data</li>
    <li>ffmpeg_7-full.dev</li>
    <li>ffmpeg_7-full.doc</li>
    <li>ffmpeg_7-full.lib</li>
    <li>ffmpeg_7-full.man</li>
  </ul>
</details>

**Review by 4evy — APPROVED** [review]
LGTM!

**Comment by toonn** [comment]
My branch fixing ffmpeg_7-full includes avfoundation, quirc and xev{d,e} patches, PR incoming later today.

Re `.so` on Darwin. It's really just an aesthetic matter TBH, Darwin doesn't actually care about the extension. Though you're right that it hints at no considerations having been made for Darwin.

**Review by arthsmn — APPROVED** [review]

**Comment by toonn** [comment]
The combination of #328420, #328422,  #328424 and #328428 should fix both ffmpeg_7 and ffmpeg_7-full on Darwin (and as a bonus potentially fix xev{d,e} support on aarch64-linux pending testing).

The changes are mostly simple but #328422 might hold things up a bit due to its complexity.

I don't mind fixing things up if this PR gets merged but it would be nice if #328424 is considered (would imply changes to this PR) so that we never need to disable AVFoundation support.

**Comment by emilazy** [comment]
I personally prefer @toonn’s changes as they avoid regressing functionality and don’t seem too complex to me.

**Comment by emilazy** [comment]
This PR was already partially incorrect because of @toonn's merged changes :( Hopefully we can get the xev{d,e} changes in soon and revert the unnecessary AVFoundation disabling.

**Pull request merged**